spi_nxp_lpspi: Fix DMA Async API

decsny opened the pull request:
Fix the ASYNC DMA API on the lpspi driver and actually make the entire driver go through that path. Rather than having a separate internally synchronous code path we can just have both APIs go through the same asynchronous path and just use wait_for_completion from spi context to implement either sync or async.

Fixes #74907

Commit:
spi_nxp_lpspi: Fix DMA Async API

Fix the ASYNC DMA API on the lpspi driver and actually make the entire
driver go through that path. Rather than having an orthogonal
internally synchronous path we can just have both APIs go through the
same asynchronous path and just use wait_for_completion from spi context
to implement either sync or async.

Also make DMA driver default y if dependency (an lpspi having dmas
property) is met.

And lpspi_wait_tx_fifo_empty can be shared between drivers.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@hakehuang do those boards pass with the DMA based driver on main? Ie is that new in this PR

decsny commented:
> > @hakehuang do those boards pass with the DMA based driver on main? Ie is that new in this PR
> 
> yes.

@hakehuang can you check this PR again, I debugged the failing kinetis and mcx platform

hakehuang commented:
> @hakehuang can you check this PR again, I debugged the failing kinetis and mcx platform

@decsny all board testing PASS, nice work!

Commit:
tests: spi_loopback: nxp: remove lpspi async=n cfg

Remove the overlay files for many of the NXP platforms, because they
were doing two things that are no longer relevant:

1) CONFIG_SPI_ASYNC=n -> LPSPI driver used to not work with
   CONFIG_SPI_ASYNC, now it does, so we should test it.
2) CONFIG_SPI_MCUX_LPSPI_DMA=y -> This is redundant with the
   testcase.yaml

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@hakehuang I just realized the test .conf has CONFIG_SPI_ASYNC=n for all the nxp boards, I don't know if this was affecting testing, but I removed these conf files

hakehuang commented:
> @hakehuang I just realized the test .conf has CONFIG_SPI_ASYNC=n for all the nxp boards, I don't know if this was affecting testing, but I removed these conf files

oh, let me re-test
